README
NAME
    App::SmokeBrew::Plugin::BINGOS - a smokebrew plugin to configure things
    like BINGOS does

VERSION
    version 0.10

SYNOPSIS
      # smokebrew.cfg

      [BINGOS]
      relay = myrelay
      port = 8080

      # then run

      $ smokebrew --plugin App::SmokeBrew::Plugin::BINGOS

DESCRIPTION
    App::SmokeBrew::Plugin::BINGOS is a App::SmokeBrew::Plugin for smokebrew
    which configures the built perl installations for CPAN Testing with
    CPANPLUS::YACSmoke and sending test reports to a metabase-relayd host
    using Test::Reporter::Transport::Socket.

    It will set up the CPANPLUS / CPANPLUS::YACSmoke base locations to be in
    the "conf" directory under the given "prefix" directory with a directory
    for each perl version.

CONFIGURATION
    This plugin requires two attributes: "relay" - the hostname or IP
    address of a metabase-relayd host and "port" - the TCP port of the
    metabase-relayd on that host.

    These attributes should be specified in the "smokebrew.cfg" file under a
    named section:

      [BINGOS]

      relay = some.host
      port = 8080

METHODS
    "configure"
        Called by smokebrew to perform the CPAN Testing configuration.

SEE ALSO
    App::SmokeBrew::Plugin

    smokebrew

    CPANPLUS

    CPANPLUS::YACSmoke

    metabase-relayd

    Test::Reporter::Transport::Socket

AUTHOR
    Chris Williams <chris@bingosnet.co.uk>

COPYRIGHT AND LICENSE
    This software is copyright (c) 2013 by Chris Williams.

    This is free software; you can redistribute it and/or modify it under
    the same terms as the Perl 5 programming language system itself.



Changes
==================================================
Changes from 2008-11-26 00:00:00 +0000 to present.
==================================================

-----------------------------------------
version 0.10 at 2013-11-25 20:01:57 +0000
-----------------------------------------

  Change: dc36832762e87941760dc2da227e5379737ac18a
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2013-11-25 20:01:57 +0000

    Move update/install of version before EUMM and M::B 

-----------------------------------------
version 0.08 at 2011-12-01 21:31:59 +0000
-----------------------------------------

  Change: 290a34e7dcd9fb675beb05140b616182839588c4
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2011-12-01 21:31:59 +0000

    Move update/install of EUMM before M::B 

-----------------------------------------
version 0.06 at 2011-01-26 09:14:10 +0000
-----------------------------------------

  Change: d9d16e0428bdcaca14edaed62c19ac30f5f18829
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2011-01-26 09:14:10 +0000

    Set the CPANPLUS configurations to 'prefer_bin' 

-----------------------------------------
version 0.04 at 2011-01-14 12:23:46 +0000
-----------------------------------------

  Change: e3ca6086a1a3cc1e35351f1b0772b14f542e40a2
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2011-01-14 12:23:46 +0000

    Bump to smokebrew prereq version to 0.28 

  Change: 7b174f5baef4bad5e02c4c0bf17d2257e54e9923
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2011-01-14 12:23:21 +0000

    Fix the cpconf script generation 

-----------------------------------------
version 0.02 at 2011-01-10 14:21:52 +0000
-----------------------------------------

  Change: e68d2eb3ccce240d3fd95bac3c1b8c8264c43bc2
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2011-01-10 14:21:52 +0000

    Add more documentation 

  Change: 7eb55530021709e70031c3940c91ac14bd0fa65f
  Author: Chris 'BinGOs' Williams <chris@bingosnet.co.uk>
  Date : 2011-01-10 13:59:31 +0000

    Go 

================
End of releases.
================


lib/App/SmokeBrew/Plugin/BINGOS.pm
package App::SmokeBrew::Plugin::BINGOS;
{
  $App::SmokeBrew::Plugin::BINGOS::VERSION = '0.10';
}

#ABSTRACT: a smokebrew plugin to configure things like BINGOS does

use strict;
use warnings;
use Moose;

extends 'App::SmokeBrew::Plugin::CPANPLUS::YACSmoke';

has 'relay' => (
  is => 'ro',
  isa => 'Str',
  required => 1,
);

has 'port' => (
  is => 'ro',
  isa => 'Str',
  required => 1,
);

sub _boxed {
my $self = shift;
return q+
BEGIN {
    use strict;
    use warnings;

    use Config;
    use FindBin;
    use File::Spec;
    use File::Spec::Unix;

    use vars qw[@RUN_TIME_INC $LIB_DIR $BUNDLE_DIR $BASE $PRIV_LIB];
    $LIB_DIR        = File::Spec->catdir( $FindBin::Bin, qw[.. lib] );
    $BUNDLE_DIR     = File::Spec->catdir( $FindBin::Bin, qw[.. inc bundle] );

    my $who     = getlogin || getpwuid($<) || $<;
    $BASE       = File::Spec->catfile(
                            $FindBin::Bin, '..', '.cpanplus', $who);
    $PRIV_LIB   = File::Spec->catfile( $BASE, 'lib' );

    @RUN_TIME_INC   = ($PRIV_LIB, @INC);
    unshift @INC, $LIB_DIR, $BUNDLE_DIR;

    $ENV{'PERL5LIB'} = join $Config{'path_sep'}, grep { defined }
                        $PRIV_LIB,              # to find the boxed config
                        #$LIB_DIR,               # the CPANPLUS libs
                        $ENV{'PERL5LIB'};       # original PERL5LIB

}

use FindBin;
use File::Find                          qw[find];
use CPANPLUS::Error;
use CPANPLUS::Configure;
use CPANPLUS::Internals::Constants;
use CPANPLUS::Internals::Utils;

{   for my $dir ( ($BUNDLE_DIR, $LIB_DIR) ) {
        my $base_re = quotemeta $dir;

        find( sub { my $file = $File::Find::name;
                return unless -e $file && -f _ && -s _;

                return if $file =~ /\._/;   # osx temp files

                $file =~ s/^$base_re(\W)?//;

                return if $INC{$file};

                my $unixfile = File::Spec::Unix->catfile(
                                    File::Spec->splitdir( $file )
                                );
                my $pm       = join '::', File::Spec->splitdir( $file );
                $pm =~ s/\.pm$//i or return;    # not a .pm file

                #return if $pm =~ /(?:IPC::Run::)|(?:File::Spec::)/;

                eval "require $pm ; 1";

                if( $@ ) {
                    push @failures, $unixfile;
                }
            }, $dir );
    }

    delete $INC{$_} for @failures;

    @INC = @RUN_TIME_INC;
}


my $ConfObj     = CPANPLUS::Configure->new;
my $Config      = CONFIG_BOXED;
my $Util        = 'CPANPLUS::Internals::Utils';
my $ConfigFile  = $ConfObj->_config_pm_to_file( $Config => $PRIV_LIB );

{   ### no base dir even, set it up
    unless( IS_DIR->( $BASE ) ) {
        $Util->_mkdir( dir => $BASE ) or die CPANPLUS::Error->stack_as_string;
    }

    unless( -e $ConfigFile ) {
        $ConfObj->set_conf( base    => $BASE );     # new base dir
        $ConfObj->set_conf( verbose => 1     );     # be verbose
        $ConfObj->set_conf( prereqs => 1     );     # install prereqs
        $ConfObj->set_conf( prefer_bin => 1 );
        $ConfObj->set_conf( prefer_makefile => 1 ); # prefer Makefile.PL because of v5.10.0
        $ConfObj->set_conf( enable_custom_sources => 0 ); # install prereqs
        $ConfObj->set_conf( hosts => + . $self->_mirrors . q+ );
        $ConfObj->set_program( sudo => undef );
        $ConfObj->save(     $Config => $PRIV_LIB ); # save the pm in that dir
    }
}

{
    $Module::Load::Conditional::CHECK_INC_HASH = 1;
    use CPANPLUS::Backend;
    my $cb = CPANPLUS::Backend->new( $ConfObj );
    my $su = $cb->selfupdate_object;

    $cb->module_tree( 'version' )->install(); # Move this here too because EUMM icky is icky :S
    $cb->module_tree( 'ExtUtils::MakeMaker' )->install(); # Move this here because icky is icky >:)
    $cb->module_tree( 'Module::Build' )->install(); # Move this here because perl-5.10.0 is icky

    $su->selfupdate( update => 'dependencies', latest => 1 );
    $cb->module_tree( $_ )->install() for
      qw(
          CPANPLUS
          File::Temp
          Compress::Raw::Bzip2
          Compress::Raw::Zlib
          Compress::Zlib
          ExtUtils::CBuilder
          ExtUtils::ParseXS
          ExtUtils::Manifest
          Log::Message::Simple
          Test::Reporter::Transport::Socket
          CPANPLUS::YACSmoke
      );
    $_->install() for map { $su->modules_for_feature( $_ ) } qw(prefer_makefile md5 storable cpantest);
}
+;
}

sub _cpconf {
  my $self = shift;
  my $cpconf = q+
use strict;
use warnings;
use Getopt::Long;
use CPANPLUS::Configure;

my $mx;
my $email;

GetOptions( 'mx=s', \$mx, 'email=s', \$email );

my $conf = CPANPLUS::Configure->new();
$conf->set_conf( verbose => 1 );
$conf->set_conf( prefer_bin => 1 );
$conf->set_conf( cpantest => 'dont_cc' );
$conf->set_conf( 'cpantest_reporter_args' =>
    {
      transport       => 'Socket',
      transport_args  => [ host => +;
  $cpconf .= sprintf( "'%s', port => '%s' ] } );", $self->relay, $self->port );
  $cpconf .= q+
$conf->set_conf( email => $email );
$conf->set_conf( makeflags => 'UNINST=1' );
$conf->set_conf( buildflags => 'uninst=1' );
$conf->set_conf( enable_custom_sources => 0 );
$conf->set_conf( show_startup_tip => 0 );
$conf->set_conf( write_install_logs => 0 );
$conf->set_conf( hosts => +;
  $cpconf .= $self->_mirrors() . ');';
  $cpconf .= q+
$conf->set_program( sudo => undef );
$conf->save();
exit 0;
+;
return $cpconf;
}

no Moose;

__PACKAGE__->meta->make_immutable;

qq[Smokin'];

__END__

=pod

=encoding UTF-8

=head1 NAME

App::SmokeBrew::Plugin::BINGOS - a smokebrew plugin to configure things like BINGOS does

=head1 VERSION

version 0.10

=head1 SYNOPSIS

  # smokebrew.cfg

  [BINGOS]
  relay = myrelay
  port = 8080

  # then run

  $ smokebrew --plugin App::SmokeBrew::Plugin::BINGOS

=head1 DESCRIPTION

App::SmokeBrew::Plugin::BINGOS is a L<App::SmokeBrew::Plugin> for L<smokebrew> which
configures the built perl installations for CPAN Testing with L<CPANPLUS::YACSmoke> and
sending test reports to a L<metabase-relayd> host using L<Test::Reporter::Transport::Socket>.

It will set up the L<CPANPLUS> / L<CPANPLUS::YACSmoke> base locations to be in the C<conf> directory
under the given C<prefix> directory with a directory for each perl version.

=head1 CONFIGURATION

This plugin requires two attributes: C<relay> - the hostname or IP address of a L<metabase-relayd>
host and C<port> - the TCP port of the L<metabase-relayd> on that host.

These attributes should be specified in the C<smokebrew.cfg> file under a named section:

  [BINGOS]

  relay = some.host
  port = 8080

=head1 METHODS

=over

=item C<configure>

Called by L<smokebrew> to perform the CPAN Testing configuration.

=back

=head1 SEE ALSO

L<App::SmokeBrew::Plugin>

L<smokebrew>

L<CPANPLUS>

L<CPANPLUS::YACSmoke>

L<metabase-relayd>

L<Test::Reporter::Transport::Socket>

=head1 AUTHOR

Chris Williams <chris@bingosnet.co.uk>

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2013 by Chris Williams.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut


Makefile.PL

use strict;
use warnings;

use 5.006000;

use ExtUtils::MakeMaker 6.30;



my %WriteMakefileArgs = (
  "ABSTRACT" => "a smokebrew plugin to configure things like BINGOS does",
  "AUTHOR" => "Chris Williams <chris\@bingosnet.co.uk>",
  "BUILD_REQUIRES" => {},
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => "6.30"
  },
  "DISTNAME" => "App-SmokeBrew-Plugin-BINGOS",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "App::SmokeBrew::Plugin::BINGOS",
  "PREREQ_PM" => {
    "App::SmokeBrew" => "0.28",
    "Moose" => "1.03"
  },
  "TEST_REQUIRES" => {
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Test::More" => 0
  },
  "VERSION" => "0.10",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "App::SmokeBrew" => "0.28",
  "File::Spec" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Moose" => "1.03",
  "Test::More" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);



